### ISSUE TITLE
Introduce ASCII-based tree printing TestExecutionListener

### ISSUE THREAD

**Issue by sbrannen** [junit-team/junit5#369]
## Status Quo

Since build tools and IDEs do not yet support the JUnit Platform, test results cannot currently be viewed in their true tree based structure. This is especially true for _dynamic tests_.
## Deliverables
- [x] Introduce an ASCII-based `TestExecutionListener` that prints the test tree to STDOUT (or a configurable output destination) once the `TestPlan` has completed.
- [x] Allow the new listener to be automatically used by the `ConsoleLauncher`.
- [x] Allow the new listener to be enabled via the JUnit Platform Gradle plugin.

**Comment by sormuras** [comment]
Although @mmerdes is assigned to this issue, I rolled my own tree printing test execution listener. See it's print out starting at line 275 here: https://travis-ci.org/sormuras/beethoven/builds/159539379 - avoiding ANSI colors and extended ASCII characters.
I'd be pleased to tidy up the [code](https://github.com/sormuras/beethoven/blob/master/src/test/java/org/junit/platform/console/tasks/ColoredPrintingTestListener.java) and file a PR as basis for further developement and tile layout tweaks.

**Comment by sbrannen** [comment]
@sormuras, I checked out the output in your build.

I personally had something more in mind of how the standard `tree` command outputs directory structures, like the following for the JUnit 5 repository:

```
.
├── CONTRIBUTING.md
├── LICENSE.md
├── README.md
├── build.gradle
├── documentation
│   ├── README.md
│   ├── bin
│   │   ├── example
│   │   │   ├── Address.class
│   │   │   ├── AssertionsDemo.class
│   │   │   ├── AssumptionsDemo.class
│   │   │   ├── DisabledClassDemo.class
│   │   │   ├── DisabledTestsDemo.class
│   │   │   ├── DisplayNameDemo.class
│   │   │   ├── DocumentationTestSuite.class
│   │   │   ├── DynamicTestsDemo$1.class
│   │   │   ├── DynamicTestsDemo.class
│   │   │   ├── Fast.class
```

So unless @mmerdes states otherwise, feel free to put together a PR if you like! We can then fine tune the output later. However you decided to implement your tree-printing `TestExecutionListener`, please make sure that it can be used standalone and that is is not dependent on existing listeners. If you feel that is not feasible, please raise your concerns here.

Cheers!

**Comment by sormuras** [comment]
@sbrannen I tried the extended ASCII characters as tree branch tiles in an older version, but I had two major issues with them:

1. They spuriously fail to display/render/print: see line 382 [here](https://travis-ci.org/sormuras/beethoven/builds/159518116). It happend on travis and locally in the IntelliJ and Eclipse console window. Flushing the output writer after each container seemed to solve that issue - but flushing feels like a bad band-aid.
2. On Windows, you have to enforce `UTF-8` as the default file encoding when executing your build/test environment. Otherwise, all extended ASCII characters are printed as `?`.

That's the reason for the nested `enum Tile` sporting different sets of tiles. Which tile set to use could be configured similar to the `ColoredPrintingTestListener#disableAnsiColors` flag.

The current implementation is just a proof-of-concept that I can use/tweak in my projects' Gradle build on-the-fly. It intentionally shadows the original class, therefore it is picked up automatically by the default `ConsoleLauncher` after a Gradle [args-hack](https://github.com/sormuras/beethoven/blob/master/build.gradle#L49). I'll move the code into its own standalone type before a PR is on its way.

**Comment by sbrannen** [comment]
Thanks for the clarification, @sormuras.

Sounds good!

**Comment by sormuras** [comment]
...something more standard [tree](https://travis-ci.org/sormuras/beethoven/builds/159677776) command like, but the spurious `???` prevail. Sigh.

```
 ├─ NameTest
 │  ├─ string()
 │  ├─ primitive()
 │  ├─ enclosed()
 │  ���─ elementInUnnamedPackage()
 │  ├─ elementInNamedPackage()
 │  ├─ equalsAndHashcode()
 │  ├─ name()
 │  ├─ cast()
 │  ├─ array()
 │  ├─ field()
 ├─ AnnotatedTest
 │  ├─ primitives()
 │  │  ├─ BOOLEAN
 │  │  ├─ BYTE
```

**Comment by mmerdes** [comment]
@sormuras 
Yes, do create a PR.

**Comment by sormuras** [comment]
Where to place the `TreePrintingTestListener` implementation?

1. `package org.junit.platform.console.tasks;` in `junit-platform-console` ... next to `ColoredPrintingTestListener`
2. `package org.junit.platform.launcher.listeners;` in `junit-platform-launcher` ... next to `SummaryGeneratingListener`

**Comment by marcphilipp** [comment]
Since I don't think it's useful outside the console (is it?) I'd put it into _ org.junit.platform.console.tasks_.

**Comment by sbrannen** [comment]
FWIW, I think we may eventually want to introduce a separate `junit-platform-reporting` module to serve as the home of XML report generation, tree-based reporting, etc.

However, for the time being I agree with @marcphilipp. Just put the tree printing listener in `org.junit.platform.console.tasks` and make it package-private.

**Comment by sbrannen** [comment]
@sormuras, I admittedly have not checked this in your current code, but... does your implementation support _dynamic tests_?

**Comment by sormuras** [comment]
@sbrannen Yes, as all dynamically registered tests executions are published via `executionStarted` and `executionFinished` as well, the tree does include those. See https://travis-ci.org/sormuras/beethoven/builds/159849098 which sports 106 static tests - and all 131 real tests are printed by the tree renderer. Will add a test to `TreePrintingTests` later.

**Comment by sbrannen** [comment]
Thanks for the link.

Are the "annotatable(...)" tests _dynamic_?

And yes, adding a dedicated test would be great.

**Comment by sormuras** [comment]
Aye. They are generated via classpath scanning at runtime. Same goes for these `primitives()` and `types()` containers some lines above. `TestFactory` sources: [AnnotatedTest](https://github.com/sormuras/beethoven/blob/master/src/test/java/com/github/sormuras/beethoven/AnnotatedTest.java)s and [AnnotatableTest](https://github.com/sormuras/beethoven/blob/master/src/test/java/com/github/sormuras/beethoven/unit/AnnotatableTest.java)s

**Comment by sbrannen** [comment]
cool. thanks for the clarification.

**Comment by sormuras** [comment]
The current iteration adds the `TreePrintingExecutionListener` to the default listeners of the console launcher - right next to the `SummaryGeneratingListener`. See forests grow at

- Linux, extended ASCII art https://travis-ci.org/junit-team/junit5/jobs/160133751
- Windows, simple ASCII art https://ci.appveyor.com/project/marcphilipp/junit5/build/179

Next task on my list: introduce an enable tree option -- and remove the forced forest from this iteration.

**Comment by sbrannen** [comment]
I'd recommend using `├──` instead of `┌─` and `-` instead of `o`.

**Comment by sormuras** [comment]
Will replace them accordingly.

What do think about the `// ` signs in the (multi-line) comments section used when printing skip reasons, stack traces and report entries?

**Comment by sbrannen** [comment]
> What do you think about the // signs in the (multi-line) comments section used when printing skip reasons, stack traces and report entries?

Hmmmm.... not sure what I think of that yet.

Other than:

1. detailed output should be an opt-in feature
1. in the very least, I would indent such information another two spaces to align underneath the test display name.

**Comment by sormuras** [comment]
Based on your last comment, I tweaked the layout to

- Windows https://ci.appveyor.com/project/marcphilipp/junit5/build/184
- Linux https://travis-ci.org/junit-team/junit5/jobs/160183009

Working on the opt-in for details feature now.

**Comment by sbrannen** [comment]
cool. thanks.

**Comment by sormuras** [comment]
I hooked into the `options.isHideDetails` switch from the console options:
```
-D, --hide-details           Hide details while tests are being executed.
                               Only show the summary and test failures.
```
Or should I introduce a new one?

**Comment by sbrannen** [comment]
I don't think we need a new switch for that.

"Details" are "details". 👍

**Comment by sbrannen** [comment]
By the way, seeing your output made me notice bug #514. 😉

**Comment by sormuras** [comment]
:bow: Saw it too, but thought it was intended.

New, close-to-final layout visible at:

1. https://travis-ci.org/junit-team/junit5/jobs/160195205
2. https://ci.appveyor.com/project/marcphilipp/junit5/build/186

Are `ReportEntry`s considered details? If yes, I'll guard them as well.

**Comment by sbrannen** [comment]
> 🙇 Saw it too, but thought it was intended.

Nope. That was definitely unintentional.

So if you notice other odd things like that, please speak up!

**Comment by sbrannen** [comment]
> Are ReportEntrys considered details? If yes, I'll guard them as well.

Yes, I consider those _details_.

**Comment by sormuras** [comment]
Consider them hidden. If `--hide-details` is true.

**Comment by sormuras** [comment]
>  Allow the new listener to be automatically used by the ConsoleLauncher.

In other words, if you start test execution via the `ConsoleLauncher`, you automatically get the `SummaryGeneratingListener` *and* the `TreePrintingTestListener` registered?
If yes, this behaviour is already implemented by this PR.

Nit, if `--hide-Details = false`, the `ColoredPrintingTestListener` is actived as well. The resulting output is ... verbose. Meaning, very detailed. May be, a dedicated switch `--hide-Tree` is needed.

> Allow the new listener to be enabled via the JUnit Platform Gradle plugin.

If the answer to the question above was _yes_ it is enabled by default. So, no need for an additional action for this deliverable, right?

This PR also publishes the `--hide-Details` switch via the Gradle plugin extension.

**Comment by sbrannen** [comment]
Note, however, that `--hide-tree` should be provided to the `ConsoleLauncher` via the Gradle plugin by default, analogous to `--hide-details`.

**Comment by sormuras** [comment]
> I agree that introducing a new --hide-details command-line option is needed.

`--hide-Tree`

**Comment by sbrannen** [comment]
Yes, I meant `--hide-tree` (with a lowercase "t").

I updated my comment accordingly.

**Comment by sbrannen** [comment]
> This PR also publishes the `--hide-details` switch via the Gradle plugin extension.

Thanks! 👍

**Comment by sbrannen** [comment]
In the `ConsoleLauncher`, as with `--hide-details`, `--hide-tree` should also default to `false`. 

In other words, the absence of `--hide-tree` as a command-line argument means implicitly that the tree is displayed.

**Comment by sormuras** [comment]
But then, if both are absent, you'll get the verbose and interleaved output of both printing listeners.

What about this decision matrix, that only one plan execution printer can be active?

hide-details | hide-tree | resulting print out
-------------|-----------|--------------------
**true** | **true** | no colors, no tree -- just **summary**
false | true | no tree -- **colors** + summary
true | false | no colors -- **tree** (simple) + summary
false | false | no colors - **tree** (with details) + summary

Or, introduce another first-level switch, like `--hide-colors` similar to `--hide-tree`?

Or, allow the user to pass a list of listener identifiers?

**Comment by sbrannen** [comment]
Hmmmm... I haven't looked into it with that level of detail.

You may be right.

But, FWIW, there is already `--disable-ansi-colors` to disable ANSI color output. So that precludes the need for `--hide-colors`.

**Comment by sormuras** [comment]
Pardon my misleading wording: `--hide-colors` was short for `--hide-potentiallyColorfulTestExectionListener` ... which could replace the `--hide-details` switch.

Maybe too much configurability is an overkill here. As IDEs will take over soon, anyway. Perhaps it's easier to merge the `Colored`- and the `TreePrintingListener` into one `DetailsPrinter` and keep the already introduced switches `--hide-details`, `--disable-ansi-colors`.

**Comment by sbrannen** [comment]
> Perhaps it's easier to merge the `Colored`- and the `TreePrintingTestListener` into one `DetailsPrinter` and keep the already introduced switches `--hide-details`, `--disable-ansi-colors`.

I think that sounds fairly reasonable.

@junit-team/junit-lambda, what do you the rest of you think about that proposal?

**Comment by sormuras** [comment]
It could, imagine colors and test identifier unique ids, look like: 
```
|  +-- DisabledTestsDemo
|  |   -  testWillBeSkipped()
             => Reason: void example.DisabledTestsDemo.testWillBeSkipped() is @Disabled
|  |   -  testWillBeExecuted()
|  |   -  testWillBeExecuted()
|  +-- DisabledTestsDemo
|  +-- A special test case
|  |   -  😱
|  |   -  😱
|  |   -  ╯°□°）╯
|  |   -  ╯°□°）╯
|  |   -  Custom test name containing spaces
|  |   -  Custom test name containing spaces
|  +-- A special test case
|  +-- DynamicTestsDemo
|  |  +-- dynamicTestsFromCollection()
|  |  |   -  1st dynamic test
|  |  |   -  1st dynamic test
|  |  |   -  1st dynamic test
|  |  |   -  2nd dynamic test
|  |  |   -  2nd dynamic test
|  |  |   -  2nd dynamic test
|  |  +-- dynamicTestsFromCollection()
```

**Comment by sormuras** [comment]
Proof-of-concept with colors and unique ids:

- https://travis-ci.org/junit-team/junit5/jobs/160420698
- https://ci.appveyor.com/project/marcphilipp/junit5/build/200

**Comment by sormuras** [comment]
Better layout, more details. https://travis-ci.org/sormuras/beethoven/builds/160647651
If you like this approach, I'll tackle skipped tests, exception stack trace and other layout improvements.

**Comment by sormuras** [comment]
Last tweak results in full-details mode:
```
Test plan execution started. Number of static tests: 125
+-- JUnit Jupiter
|  
|  +-- ListableTests
|  |  
|  |  +-- escapeCharacter()
|  |  |     tags: []
|  |  |       id: [engine:junit-jupiter]/[class:com.github.sormuras.beethoven.ListableTests]/[method:escapeCharacter()]
|  |  |   parent: [engine:junit-jupiter]/[class:com.github.sormuras.beethoven.ListableTests]
|  |  |   source: JavaMethodSource [javaClass = 'com.github.sormuras.beethoven.ListableTests', javaMethodName = 'escapeCharacter', javaMethodParameterTypes = '']
|  |  | duration: 14 ms
|  |  = SUCCESSFUL
|  |  
|  |  +-- escapeString()
|  |  |     tags: []
...
|  ListableTests took 115 ms and was SUCCESSFUL
|  
|  +-- NameTests
|  |  
|  |  +-- string()
|  |  |     tags: []
...
|  |  
|  |  +-- lastModified()
|  |  |     tags: []
|  |  |       id: [engine:junit-jupiter]/[class:com.github.sormuras.beethoven.CompilationTests]/[method:lastModified()]
|  |  |   parent: [engine:junit-jupiter]/[class:com.github.sormuras.beethoven.CompilationTests]
|  |  |   source: JavaMethodSource [javaClass = 'com.github.sormuras.beethoven.CompilationTests', javaMethodName = 'lastModified', javaMethodParameterTypes = '']
|  |  | duration: 2 ms
|  |  = SUCCESSFUL
|  |  
|  CompilationTests took 759 ms and was SUCCESSFUL
|  
JUnit Jupiter took 4692 ms and was SUCCESSFUL
```
Entire output with colors is here: https://travis-ci.org/sormuras/beethoven/builds/160805090

**Comment by sormuras** [comment]
@sbrannen Any decision taken regarding the `DetailsPrinter`?

The current p-o-c version looks like:

* Simple tree https://travis-ci.org/sormuras/beethoven/builds/161607359
* Verbose tree https://travis-ci.org/sormuras/beethoven/builds/161608341

**Comment by sormuras** [comment]
In #511 I replaced the simple `--hide-details` switch with an enum called [DetailMode](https://github.com/sormuras/junit5/blob/f711d3713c8e0702920229becb373b041382fdf0/junit-platform-console/src/main/java/org/junit/platform/console/options/DetailMode.java). It provides an `HIDDEN`, `FLAT` and two tree variants. With this, the user can select the desired detail mode by naming it - no need for the decision table discussed above.

**Comment by sormuras** [comment]
@sbrannen Closing #521 and opening #524 didn't transfer the labels you attached. What do you guys think about the `-details=[hidden, flat, tree, verbose]` switch?

**Comment by marcphilipp** [comment]
_in progress_ due to #524

**Comment by sbrannen** [comment]
This has been addressed in `master` in a64e354fa285ecd9297521ca327d9fbed34c21c9.
